src/main/java/com/conveyal/r5/profile/PerTargetPropagater.java

```diff
@@ -0,0 +1,107 @@
+package com.conveyal.r5.profile;
+
+import com.conveyal.r5.streets.LinkedPointSet;
+import gnu.trove.map.TIntIntMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+
+/**
+ * This is similar to the Propagater class, but instead of the reducer function being called with each iteration,
+ * it is called with the travel times to a particular target for all iterations.
+ *
+ * The reducer will be called with the target index, and a boolean array of whether that target is reachable at each iteration.
+ */
+public class PerTargetPropagater {
+    private static final Logger LOG = LoggerFactory.getLogger(PerTargetPropagater.class);
+
+    /** Times at transit stops for each iteration */
+    public final int[][] travelTimesToStopsEachIteration;
+
+    /** Times at targets using the street network */
+    public final int[] nonTransferTravelTimesToTargets;
+
+    /** The maximum allowable travel time in this regional analysis, to avoid propagating things that won't be useful */
+    public final int cutoffSeconds;
+
+    /** The linked targets */
+    public final LinkedPointSet targets;
+
+    /** the profilerequest (used for walk speed etc.) */
+    public final ProfileRequest request;
+
+    public PerTargetPropagater (int[][] travelTimesToStopsEachIteration, int[] nonTransferTravelTimesToTargets, LinkedPointSet targets, ProfileRequest request, int cutoffSeconds) {
+        this.travelTimesToStopsEachIteration = travelTimesToStopsEachIteration;
+        this.nonTransferTravelTimesToTargets = nonTransferTravelTimesToTargets;
+        this.targets = targets;
+        this.request = request;
+        this.cutoffSeconds = cutoffSeconds;
+    }
+
+    public void propagate (Reducer reducer) {
+        targets.makePointToStopDistanceTablesIfNeeded();
+
+        long startTimeMillis = System.currentTimeMillis();
+        // avoid float math in loop below
+        // float math was previously observed to slow down this loop, however it's debatable whether that was due to
+        // casts, the operations themselves, or the fact that the operations were being completed with doubles rather
+        // than floats.
+        int speedMillimetersPerSecond = (int) (request.walkSpeed * 1000);
+
+        for (int targetIdx = 0; targetIdx < targets.size(); targetIdx++) {
+            boolean[] perIterationResults = new boolean[travelTimesToStopsEachIteration.length];
+            // fill with non-transit time as upper bound on travel time
+            Arrays.fill(perIterationResults, nonTransferTravelTimesToTargets[targetIdx] < cutoffSeconds);
+
+            if (perIterationResults[0]) {
+                // reachable on foot, no need to even check if it's reachable via transit
+                reducer.accept(targetIdx, perIterationResults);
+                continue;
+            }
+
+            TIntIntMap pointToStopDistanceTable = targets.pointToStopDistanceTables.get(targetIdx);
+
+            // effectively final nonsense
```

Comment:
Elaborate in comment: Java can only close lambdas over effectively final variables, and an array of one value meets that criterion while still allowing the value to be modified.

It's too bad Trove maps don't have a method that allows you to map-reduce all their entries (pass each pair of entries through a function and collect the results like streams).